src/rewriting.jl

```diff
@@ -0,0 +1,82 @@
+"""
+    AbstractRewritingSystem{T} <: AbstractVector{T}
+Abstract type representing words over an Alphabet.
+
+`AbstractRewritingSystem` as such has its meaning only in the contex of an Alphabet.
+The subtypes of `AbstractRewritingSystem{T}` need to implement the following
+methods which constitute `AbstractRewritingSystem` interface:
+ * `Base.push!`/`Base.pushfirst!`: appending a single rule at the end/beginning
+ * `Base.append!`/`Base.prepend!`: appending a another system at the end/beginning,
+ * `Base.insert!`: inserting a single rule at a given place
+ * `Base.delateat!`: delating rules at given positions
+ * `Base.empty!`: delating all the rules
+ * full iteration protocol for `AbstractArray`s, returning pairs of Words (rules)
+ * `length` the length of word as written in the alphabet.
+"""
+
+abstract type AbstractRewritingSystem{T} <: AbstractVector{T} end
+
+"""
+    RewritingSystem{T} <: AbstractRewritingSystem{T}
+RewritingSystem written as a list of pairs of Words (left side => right side)
+"""
+struct RewritingSystem{T} <: AbstractRewritingSystem{T}
+    rwrules::Vector{T}
```

Comment:
This has pros and cons:
- in general `RewritingSystem` needs no `Ordering` - it is just a set of rules (I think of it as the "~" part  of "< ... |  ~  >"  monoid/group);
- as to `test1!` and `overlap1!` - at that stage I wanted to be consistent with Sims' book. After all I believe that the real thing will start with `kbs2` (it has substantial differences) - names are just a final polish. Keeping things consistent with the book has some advantage at that stage: it allows to easily look up things - seems trivial, but ...
- `getirrsubsys` - this is probably only important in `kbs1`, later there will be no such thing as we will remove rules at the earlier stage ('active'/'inactive' in the book). BTW: starting from `kbs2` it may be useful to make `RewritingSystem` possess a dictionary of rules instead of a list of rules (as the left sides will be always unique according to the book).

Thus answering the first (implicit) question: I would opt for not having too much in `RewritingSystem` at the moment - it will make things more rigid and I am not sure it is a stage at which we want things to be too rigid.

Reply:
that is true, but only if we always start from empty `RewritingSystem`. Otherwise (e.g. resuming `knuthbendix` after increasing the number of rules, or just running again with some time-budget) we need to check if the existing rules in `rws` are consistent with the given ordering.

I'm fine with the names for now, this was just to spark some discussion;

the ordering can stay outside of `RewritingSystem` for now (before we dig into `kbs2` and automata), but I guess it will be much easier to reason about the functions (`knuthbendix` in particular) when ordering is stored inside and cannot be changed in the middle of lifespan of an `rws`

Reply:
OK, now I see the argument for including the `Ordering` into the `RewritingSystem` and will  implement it.